File: course_3/module_2/examples/kruskal_mst.go
package main

import (
	"fmt"
	"sort"
)

// Edge represents an edge in the graph.
type Edge struct {
	U, V   int     // The vertices connected by the edge.
	Weight float64 // The weight of the edge.
}

// Graph represents a graph with vertices and edges.
type Graph struct {
	Vertices int    // Number of vertices in the graph.
	Edges    []Edge // List of edges in the graph.
}

// AddEdge adds an edge to the graph.
func (g *Graph) AddEdge(u, v int, weight float64) {
	if u < 0 || v < 0 || u >= g.Vertices || v >= g.Vertices {
		panic(fmt.Sprintf("Vertices %d and %d must be between 0 and %d", u, v, g.Vertices-1))
	}
	g.Edges = append(g.Edges, Edge{U: u, V: v, Weight: weight})
}

// UnionFind is a structure for managing disjoint sets.
type UnionFind struct {
	Parent []int // Parent pointers for each element.
	Rank   []int // Rank (depth) of each tree.
}

// NewUnionFind creates a new UnionFind for the given size.
func NewUnionFind(size int) *UnionFind {
	parent := make([]int, size)
	rank := make([]int, size)
	for i := range parent {
		parent[i] = i
	}
	return &UnionFind{Parent: parent, Rank: rank}
}

// Find finds the root of the set containing x, using path compression.
func (uf *UnionFind) Find(x int) int {
	if uf.Parent[x] != x {
		uf.Parent[x] = uf.Find(uf.Parent[x]) // Path compression.
	}
	return uf.Parent[x]
}

// Union unites the sets containing x and y, using union by rank.
func (uf *UnionFind) Union(x, y int) {
	rootX := uf.Find(x)
	rootY := uf.Find(y)

	if rootX != rootY {
		if uf.Rank[rootX] > uf.Rank[rootY] {
			uf.Parent[rootY] = rootX
		} else if uf.Rank[rootX] < uf.Rank[rootY] {
			uf.Parent[rootX] = rootY
		} else {
			uf.Parent[rootY] = rootX
			uf.Rank[rootX]++
		}
	}
}

// KruskalMST computes the Minimum Spanning Tree using Kruskal's algorithm.
func (g *Graph) KruskalMST() ([]Edge, float64) {
	// Sort edges by weight.
	sort.Slice(g.Edges, func(i, j int) bool {
		return g.Edges[i].Weight < g.Edges[j].Weight
	})

	uf := NewUnionFind(g.Vertices)

	var mst []Edge
	var totalWeight float64

	// Process edges in order of increasing weight.
	for _, edge := range g.Edges {
		if uf.Find(edge.U) != uf.Find(edge.V) {
			uf.Union(edge.U, edge.V)
			mst = append(mst, edge)
			totalWeight += edge.Weight
		}
	}

	return mst, totalWeight
}

func main() {
	// Create a graph with 5 vertices.
	graph := &Graph{Vertices: 5}

	// Add edges to the graph.
	graph.AddEdge(0, 1, 1)
	graph.AddEdge(0, 2, 3)
	graph.AddEdge(1, 2, 2)
	graph.AddEdge(1, 3, 5)
	graph.AddEdge(2, 3, 4)
	graph.AddEdge(3, 4, 6)

	// Compute the Minimum Spanning Tree.
	mst, totalWeight := graph.KruskalMST()

	// Output the results.
	fmt.Println("Edges in the MST:")
	for _, edge := range mst {
		fmt.Printf("Edge(%d, %d, %.2f)\n", edge.U, edge.V, edge.Weight)
	}
	fmt.Printf("Total weight of the MST: %.2f\n", totalWeight)
}

File: course_3/module_2/examples/kruskal_mst.py
class Edge:
    def __init__(self, u, v, weight):
        self.u = u  # The starting vertex of the edge.
        self.v = v  # The ending vertex of the edge.
        self.weight = weight  # The weight of the edge.

    def __repr__(self):
        return f"Edge({self.u}, {self.v}, {self.weight})"


class Graph:
    def __init__(self, vertices):
        self.vertices = vertices  # The number of vertices in the graph.
        self.edges = []  # The list of edges in the graph.

    def add_edge(self, u, v, weight):
        # Add an edge between vertex u and vertex v with a given weight.
        if u < 0 or v < 0 or u >= self.vertices or v >= self.vertices:
            raise ValueError(f"Vertices {u} and {v} must be between 0 and {self.vertices - 1}.")
        self.edges.append(Edge(u, v, weight))

    def kruskal_mst(self):
        # Sort edges by weight.
        sorted_edges = sorted(self.edges, key=lambda edge: edge.weight)

        # Create a Union-Find instance to manage connected components.
        uf = UnionFind(self.vertices)

        mst = []  # The edges in the minimum spanning tree.
        total_weight = 0  # The total weight of the MST.

        for edge in sorted_edges:
            # Check if the edge creates a cycle.
            if uf.find(edge.u) != uf.find(edge.v):
                uf.union(edge.u, edge.v)
                mst.append(edge)
                total_weight += edge.weight

        return mst, total_weight


class UnionFind:
    def __init__(self, size):
        self.parent = list(range(size))  # Parent pointers for each element.
        self.rank = [0] * size  # The rank (tree depth) for each element.

    def find(self, x):
        # Find the root of the set containing x with path compression.
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # Path compression.
        return self.parent[x]

    def union(self, x, y):
        # Union two sets by rank.
        root_x = self.find(x)
        root_y = self.find(y)

        if root_x != root_y:
            if self.rank[root_x] > self.rank[root_y]:
                self.parent[root_y] = root_x
            elif self.rank[root_x] < self.rank[root_y]:
                self.parent[root_x] = root_y
            else:
                self.parent[root_y] = root_x
                self.rank[root_x] += 1


# Example usage.
if __name__ == "__main__":
    # Create a graph with 5 vertices.
    graph = Graph(5)

    # Add edges (u, v, weight).
    graph.add_edge(0, 1, 1)
    graph.add_edge(0, 2, 3)
    graph.add_edge(1, 2, 2)
    graph.add_edge(1, 3, 5)
    graph.add_edge(2, 3, 4)
    graph.add_edge(3, 4, 6)

    # Find the minimum spanning tree.
    mst, total_weight = graph.kruskal_mst()

    print("Edges in the MST:")
    for edge in mst:
        print(edge)
    print(f"Total weight of the MST: {total_weight}")
